### PR TITLE
break out scala interpreter initialization

### PR DESCRIPTION
**Pull request by brad-kaiser**
Here are the changes to break out the initialization of the scala interpreter. I'm a little worried about the Option[Future[ScalaInterpreter]] floating around in KernelBootstraps, but I'm not sure of a way to get around it. I was going to put a ready field in the ScalaInterpreter that you could await on, but that's problematic because ScalaInterpreters can be reset which would make that messy and potentially incorrect. I also thought about using an OptionT, but I don't think that would really get me anything. 

This pull request is just against my own repo, Once you guys review I'll make a real PR

### CODE REVIEW

File: scala-interpreter/src/main/scala/org/apache/toree/kernel/interpreter/scala/ScalaInterpreter.scala [hunk 1]
```diff
@@ -99,6 +99,33 @@ class ScalaInterpreter(private val config:Config = ConfigFactory.load) extends I
...
+   *
+   * @return this partially initialized scalaInterpreter
+   */
+  def startInit(): ScalaInterpreter = {
+    import scala.collection.JavaConverters._
+    val args = config.getStringList("interpreter_args").asScala.toList
+    settings = newSettings(args)
+    settings = appendClassPath(settings)
+
+    start()
```
Comment (cin) [inline scala-interpreter/src/main/scala/org/apache/toree/kernel/interpreter/scala/ScalaInterpreter.scala:113]: `start` is being called in `init` too. Is this potentially problematic? If this is the same `start` from `ScalaInterpreterSpecific`, won't this throw an exception due to the require? Actually, I think I see what's going on. It appears you don't call the init method anymore.

File: scala-interpreter/src/main/scala-2.11/org/apache/toree/kernel/interpreter/scala/ScalaInterpreterSpecific.scala [hunk 2]
```diff
@@ -258,14 +258,13 @@ trait ScalaInterpreterSpecific extends SettingsProducerLike { this: ScalaInterpr
    * Starts the interpreter, initializing any internal state.
    * @return A reference to the interpreter
    */
-  override def start(): Interpreter = {
+  override def start(): ScalaInterpreter = {
```
Comment (cin) [inline scala-interpreter/src/main/scala-2.11/org/apache/toree/kernel/interpreter/scala/ScalaInterpreterSpecific.scala:261]: Was the change in return type intentional? Doesn't this break the override?

**Review by cin — APPROVED** [review]
overall LGTM. Using Future[Option[ScalaInterpreter]] would make the futures more composable but what you have is fine IMO. I'm not even sure it'd be easier to read. Interesting maybeEventual variable names...

File: scala-interpreter/src/main/scala/org/apache/toree/kernel/interpreter/scala/ScalaInterpreter.scala (hunk 1 above)
Comment (brad-kaiser) [reply scala-interpreter/src/main/scala/org/apache/toree/kernel/interpreter/scala/ScalaInterpreter.scala:113]: yep.

File: scala-interpreter/src/main/scala-2.11/org/apache/toree/kernel/interpreter/scala/ScalaInterpreterSpecific.scala (hunk 2 above)
Comment (brad-kaiser) [reply scala-interpreter/src/main/scala-2.11/org/apache/toree/kernel/interpreter/scala/ScalaInterpreterSpecific.scala:261]: It doesn't and its kind of handy in my startInit method.

**Comment by brad-kaiser** [comment]
The maybeEventually names were suggested by Intellij and I liked it. I did Option[Future[ScalaInterpreter]] so that we won't use a thread or anything in the case of a None.

File: kernel/src/main/scala/org/apache/toree/boot/layer/ComponentInitialization.scala [hunk 3]
```diff
@@ -36,22 +37,39 @@ import org.apache.toree.utils.{LogLike, FileUtils}
...
+   * @param maybeEventualScalaInterp  a future that will hold the Scala Interpreter when its ready
    */
   def initializeComponents(
-    config: Config, actorLoader: ActorLoader
+      config: Config,
+      actorLoader: ActorLoader,
+      interpreterManager: InterpreterManager,
+      maybeEventualScalaInterp: Option[Future[ScalaInterpreter]]
   ): (CommStorage, CommRegistrar, CommManager, Interpreter,
     Kernel, DependencyDownloader, MagicManager, PluginManager,
```
Comment (ecurtin) [inline kernel/src/main/scala/org/apache/toree/boot/layer/ComponentInitialization.scala:71]: I know this is not your code, but this is a LOT of stuff to return from one method. Might be better organized as member variables in one case class.

**Review by ecurtin — APPROVED** [review]
LGTM. Have you had a chance to manually run a test where you kill off stuff in the background before its done initializing?

**Comment by brad-kaiser** [comment]
@ecurtin like a test where I killed off other program components? I haven't really thought about that.

File: kernel/src/main/scala/org/apache/toree/boot/layer/ComponentInitialization.scala (hunk 3 above)
Comment (brad-kaiser) [reply kernel/src/main/scala/org/apache/toree/boot/layer/ComponentInitialization.scala:71]: agreed

**Pull request closed without merging**